libs/client-api/src/v2/actor.rs

```diff
@@ -0,0 +1,802 @@
+      if let Some(inner) = weak_inner.upgrade() {
+        let rid: ActionSource = tx
+          .origin()
+          .and_then(|origin| Rid::from_bytes(origin.as_ref()).ok())
+          .into();
+        tracing::trace!(
+          "[{}] emit collab update {:?} {:#?} ",
+          client_id,
+          rid,
+          Update::decode_v1(&e.update).unwrap()
+        );
+        if let ActionSource::Remote(rid) = rid {
+          tracing::trace!("[{}] {} received collab from remote", client_id, object_id);
+          last_message_id.rcu(|old| {
+            if rid > **old {
+              Arc::new(rid)
+            } else {
+              old.clone()
+            }
+          });
+        } else {
+          sync_state.set_sync_state(SyncState::Syncing);
+        }
+        inner.publish_update(object_id, collab_type, rid, e.update.clone());
+      }
+    })?;
+    let weak_inner = Arc::downgrade(actor);
+    awareness.on_change_with("af", move |awareness, e, origin| {
+      if let Some(inner) = weak_inner.upgrade() {
+        if origin.map(|o| o.as_ref()) != Some(Self::REMOTE_ORIGIN.as_bytes()) {
+          match awareness.update_with_clients(e.all_changes()) {
+            Ok(update) => inner.publish_awareness(object_id, collab_type, update),
+            Err(err) => tracing::error!(
+              "[{}] failed to prepare awareness update for {}: {}",
+              client_id,
+              object_id,
+              err
+            ),
+          }
+        }
+      }
+    });
+    actor.publish_manifest(collab, collab_type);
+    actor.publish_awareness(object_id, collab_type, awareness.update()?);
+    actor.cache.insert(object_id, Arc::downgrade(collab_ref));
+    Ok(())
+  }
+
+  fn set_connection_status(&self, status: ConnectionStatus) {
+    let _ = self.status_tx.send(status);
+  }
+
+  async fn ping(&self) -> anyhow::Result<()> {
+    if let Some(conn) = self.ws_sink() {
+      let mut lock = conn.lock().await;
+      lock.send(Message::Ping(Vec::new())).await?;
+      lock.flush().await?;
+    }
+    Ok(())
+  }
+
+  async fn actor_loop(
+    weak_ref: Weak<WorkspaceControllerActor>,
+    mut receiver: ChannelReceiverCompactor,
+  ) {
+    let mut keep_alive = tokio::time::interval(Self::PING_INTERVAL);
+    keep_alive.set_missed_tick_behavior(MissedTickBehavior::Delay);
+    loop {
+      select! {
+        action = receiver.recv() => {
+          match action {
+            None => break,
+            Some(action) => {
+              if let Some(actor) = weak_ref.upgrade() {
+                Self::handle_action(&actor, action).await;
+              }
+            }
+          }
+        }
+        _ = keep_alive.tick() => {
+          // we didn't receive any message for some time, so we send a ping to keep connection alive
+          let actor = match weak_ref.upgrade() {
+            Some(actor) => actor,
+            None => break, // controller dropped
+          };
+          if let Err(err) = actor.ping().await {
```

Comment:
If I understand correctly, each time a ping fails we set the status to “Disconnected.” If the client immediately attempts to reconnect upon seeing that status, it’s very easy to trigger repeated reconnects, which will cause the server to receive numerous Join messages.

Reply:
Failure to send a ping, means that the connection is already broken, so it cannot be used to send any more messages reliably anyway.

The exponentiall backoff reconnect strategy should be responsibility of the one calling the reconnects, which would mean the client.